<a href="https://colab.research.google.com/github/UthamanA/Playstore-App-review-EDA---Capstone-1/blob/main/Play_store_app_review.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Google Play Store Apps Exploratory Data Analysis (EDA)

##Introduction
Google Play Store or formerly Android Market, is a digital distribution service developed and operated by Google. It is an official apps store that provides variety content such as apps, books, magazines, music, movies and television programs. It serves an as platform to allow users with 'Google certified' Android operating system devices to donwload applications developed and published on the platform either with a charge or free of cost. With the rapidly growth of Android devices and apps, it would be interesting to perform data analysis on the data to obtain valuable insights.

The dataset that is going to be used is 'Google Play Store Apps' from Kaggle. It contains 10k of web scraped Play Store apps data for analysing the Android market. The tools that are going to be used for this EDA would be numpy, pandas, matplotlib and seaborn which I have learnt from the course.

##Data Preparation and Cleaning
In this section, we will be loading the Google Store Apps data stored in csv using pandas which is a fast and powerful python library for data analysis and easy data manipulation in pandas DataFrame object. It is usually used for working with tabular data (e.g data in spreadsheet) in various formats such as CSV, Excel spreadsheets, HTML tables, JSON etc. We will then perform some data preparation and also cleaning on it.

In [2]:

# import the necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline 
# allow matplotlib to plot inline with frontends like Jupyter

In [4]:
# load the apps and reviews data into pandas dataframe
apps_df = pd.read_csv('/content/drive/MyDrive/Capstone Project/capstone pro 1 - EDA/Play store app/Play Store Data.csv')

In [5]:
# look at the first 10 records in the apps dataframe
apps_df.head(10)

App  ...   Android Ver
0     Photo Editor & Candy Camera & Grid & ScrapBook  ...  4.0.3 and up
1                                Coloring book moana  ...  4.0.3 and up
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ...  4.0.3 and up
3                              Sketch - Draw & Paint  ...    4.2 and up
4              Pixel Draw - Number Art Coloring Book  ...    4.4 and up
5                         Paper flowers instructions  ...    2.3 and up
6            Smoke Effect Photo Maker - Smoke Editor  ...  4.0.3 and up
7                                   Infinite Painter  ...    4.2 and up
8                               Garden Coloring Book  ...    3.0 and up
9                      Kids Paint Free - Drawing Fun  ...  4.0.3 and up

[10 rows x 13 columns]

In [6]:
# look at the random 10 records in the apps dataframe
apps_df.sample(10)

App  ...         Android Ver
844   HelloTalk — Chat, Speak & Learn Foreign Languages  ...          4.1 and up
618                                      FREE LIVE TALK  ...        4.0.3 and up
3069                        FotMob - Live Soccer Scores  ...  Varies with device
2923                      FilterGrid - Cam&Photo Editor  ...          4.0 and up
1485  N1.RU - Real estate: apartments, new buildings...  ...        4.0.3 and up
1554                     Ayyush wih wih new without Net  ...          2.3 and up
6318                          BJ Grand Salon Mobile App  ...          5.0 and up
2189                                  Avokiddo Emotions  ...        4.0.3 and up
4236                              I'm Ping Pong King :)  ...          4.1 and up
5076                50000 Free eBooks & Free AudioBooks  ...          4.4 and up

[10 rows x 13 columns]

##Description of App Dataset columns
1.App : The name of the app

2.Category : The category of the app

3.Rating : The rating of the app in the Play Store

4.Reviews : The number of reviews of the app

5.Size : The size of the app

6.Install : The number of installs of the app

7.Type : The type of the app (Free/Paid)

8.The price of the app (0 if it is Free)

9.Content Rating :The appropiate target audience of the app
10.Genres: The genre of the app

11.Last Updated : The date when the app was last updated

12.Current Ver : The current version of the app

13.Android Ver : The minimum Android version required to run the app

In [7]:
# type of Category
apps_df['Category'].unique()

array(['ART_AND_DESIGN', 'AUTO_AND_VEHICLES', 'BEAUTY',
       'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMICS', 'COMMUNICATION',
       'DATING', 'EDUCATION', 'ENTERTAINMENT', 'EVENTS', 'FINANCE',
       'FOOD_AND_DRINK', 'HEALTH_AND_FITNESS', 'HOUSE_AND_HOME',
       'LIBRARIES_AND_DEMO', 'LIFESTYLE', 'GAME', 'FAMILY', 'MEDICAL',
       'SOCIAL', 'SHOPPING', 'PHOTOGRAPHY', 'SPORTS', 'TRAVEL_AND_LOCAL',
       'TOOLS', 'PERSONALIZATION', 'PRODUCTIVITY', 'PARENTING', 'WEATHER',
       'VIDEO_PLAYERS', 'NEWS_AND_MAGAZINES', 'MAPS_AND_NAVIGATION',
       '1.9'], dtype=object)

In [8]:
# type of Type
apps_df['Type'].unique()

array(['Free', 'Paid', nan, '0'], dtype=object)

In [9]:
# type of Content Rating
apps_df['Content Rating'].unique()

array(['Everyone', 'Teen', 'Everyone 10+', 'Mature 17+',
       'Adults only 18+', 'Unrated', nan], dtype=object)

In [10]:
# type of Genres
apps_df['Genres'].unique()

array(['Art & Design', 'Art & Design;Pretend Play',
       'Art & Design;Creativity', 'Art & Design;Action & Adventure',
       'Auto & Vehicles', 'Beauty', 'Books & Reference', 'Business',
       'Comics', 'Comics;Creativity', 'Communication', 'Dating',
       'Education;Education', 'Education', 'Education;Creativity',
       'Education;Music & Video', 'Education;Action & Adventure',
       'Education;Pretend Play', 'Education;Brain Games', 'Entertainment',
       'Entertainment;Music & Video', 'Entertainment;Brain Games',
       'Entertainment;Creativity', 'Events', 'Finance', 'Food & Drink',
       'Health & Fitness', 'House & Home', 'Libraries & Demo',
       'Lifestyle', 'Lifestyle;Pretend Play',
       'Adventure;Action & Adventure', 'Arcade', 'Casual', 'Card',
       'Casual;Pretend Play', 'Action', 'Strategy', 'Puzzle', 'Sports',
       'Music', 'Word', 'Racing', 'Casual;Creativity',
       'Casual;Action & Adventure', 'Simulation', 'Adventure', 'Board',
       'Trivia', 'Role 

In [11]:
# type of Genres
apps_df['Genres'].unique()

array(['Art & Design', 'Art & Design;Pretend Play',
       'Art & Design;Creativity', 'Art & Design;Action & Adventure',
       'Auto & Vehicles', 'Beauty', 'Books & Reference', 'Business',
       'Comics', 'Comics;Creativity', 'Communication', 'Dating',
       'Education;Education', 'Education', 'Education;Creativity',
       'Education;Music & Video', 'Education;Action & Adventure',
       'Education;Pretend Play', 'Education;Brain Games', 'Entertainment',
       'Entertainment;Music & Video', 'Entertainment;Brain Games',
       'Entertainment;Creativity', 'Events', 'Finance', 'Food & Drink',
       'Health & Fitness', 'House & Home', 'Libraries & Demo',
       'Lifestyle', 'Lifestyle;Pretend Play',
       'Adventure;Action & Adventure', 'Arcade', 'Casual', 'Card',
       'Casual;Pretend Play', 'Action', 'Strategy', 'Puzzle', 'Sports',
       'Music', 'Word', 'Racing', 'Casual;Creativity',
       'Casual;Action & Adventure', 'Simulation', 'Adventure', 'Board',
       'Trivia', 'Role 

##By diagnosing the data frame, we know that:

1.There are 13 columns of properties with 10841 rows of data.

2.Column 'Reviews', 'Size', 'Installs' and 'Price' are in the type of 'object'

3.Values of column 'Size' are strings representing size in 'M' as Megabytes, 'k' as kilobytes and also 'Varies with devices'.

4.Values of column 'Installs' are strings representing install amount with symbols such as ',' and '+'.

5.Values of column 'Price' are strings representing price with symbol '$'.

Hence, we will need to do some data cleaning.

##Some Data Cleaning

In [12]:
# 1) clean the 'Reviews' data and change the type 'object' to 'float'
reviews = [i for i in apps_df['Reviews']]

def clean_reviews(reviews_list):
    """
    As 'M' has been found the in reviews data, this function
    replace it with million
    """
    cleaned_data = []
    for review in reviews_list:
        if 'M' in review:
            review = review.replace('M', '')
            review = float(review) * 1000000  # 1M = 1,000,000
        cleaned_data.append(review)
    return cleaned_data

apps_df['Reviews'] = clean_reviews(reviews)
apps_df['Reviews'] = apps_df['Reviews'].astype(float)

In [13]:
# 2) clean the 'Size' data and change the type 'object' to 'float'

# found value with '1,000+' in one of record, remove it from data_frame as uncertain whether it is 'M' or 'k'
index = apps_df[apps_df['Size'] == '1,000+'].index
apps_df.drop(axis=0, inplace=True, index=index)

sizes = [i for i in apps_df['Size']]

def clean_sizes(sizes_list):
    """
    As sizes are represented in 'M' and 'k', we remove 'M'
    and convert 'k'/kilobytes into megabytes
    """
    cleaned_data = []
    for size in sizes_list:
        if 'M' in size:
            size = size.replace('M', '')
            size = float(size)
        elif 'k' in size:
            size = size.replace('k', '')
            size = float(size)
            size = size/1024  # 1 megabyte = 1024 kilobytes
        # representing 'Varies with device' with value 0
        elif 'Varies with device' in size:
            size = float(0)
        cleaned_data.append(size)
    return cleaned_data

apps_df['Size'] = clean_sizes(sizes)
apps_df['Size'] = apps_df['Size'].astype(float)

In [14]:
# 3) clean the 'Installs' data and change the type 'object' to 'float'
installs = [i for i in apps_df['Installs']]

def clean_installs(installs_list):
    cleaned_data = []
    for install in installs_list:
        if ',' in install:
            install = install.replace(',', '')
        if '+' in install:
            install = install.replace('+', '')
        install = int(install)
        cleaned_data.append(install)
    return cleaned_data
        
apps_df['Installs'] = clean_installs(installs)
apps_df['Installs'] = apps_df['Installs'].astype(float)

In [15]:
# 4) clean the 'Price' data and change the type 'object' to 'float'
prices = [i for i in apps_df['Price']]

def clean_prices(prices_list):
    cleaned_data = []
    for price in prices_list:
        if '$' in price:
            price = price.replace('$', '')
        cleaned_data.append(price)
    return cleaned_data

apps_df['Price'] = clean_prices(prices)
apps_df['Price'] = apps_df['Price'].astype(float)

In [16]:
# look at the random 10 records in the apps dataframe to verify the cleaned columns
apps_df.sample(10)

App  ...         Android Ver
4467                P Launcher Theme For Android™ 9.0  ...          4.0 and up
4671                           V LIVE - Star Live App  ...  Varies with device
3344                         Solo Locker (DIY Locker)  ...          4.2 and up
10585                                        FL House  ...  Varies with device
4058                         Z Ringtones PREMIUM 2018  ...        4.0.3 and up
1560                               New Ringtones 2018  ...          4.0 and up
4208                Alphabet "H" Passcode Lock Screen  ...          4.1 and up
2932                        Pixlr – Free Photo Editor  ...        4.0.3 and up
5349   Muslim Pro - Prayer Times, Azan, Quran & Qibla  ...  Varies with device
10007                                  EW Neuron Scan  ...          5.0 and up

[10 rows x 13 columns]

In [17]:
# check on null values
apps_df.isna().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       0
Genres               0
Last Updated         0
Current Ver          8
Android Ver          2
dtype: int64

Here, we realized that there are 1474 rows having null values under column 'Rating'. Hence, we decided to replace the null values with median of overall 'Rating' values.

In [18]:
def replace_with_median(series):
    """
    Given a series, replace the rows with null values 
    with median values
    """
    return series.fillna(series.median())

apps_df['Rating'] = apps_df['Rating'].transform(replace_with_median)
apps_df['Rating'] = apps_df['Rating'].astype(float)

In [19]:
# remove the record where 'Type' is having null value
index = apps_df[apps_df['Type'].isna()].index
apps_df.drop(axis=0, inplace=True, index=index)

In [20]:
# check on null values
apps_df.isna().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       8
Android Ver       2
dtype: int64

In [21]:
# grouping the data starting with App and Reviews
apps_df = apps_df.groupby(['App', 'Reviews', 'Category', 'Rating', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 
                           'Last Updated', 'Current Ver', 'Android Ver'], as_index=False)
# reassign Installs values with their mean
apps_df = apps_df['Installs'].mean()
# sort the dataframe by Reviews descendingly
apps_df.sort_values(by='Reviews', ascending=False, inplace=True)
# drop duplicate rows based on App 
apps_df.drop_duplicates(subset=['App'], inplace=True)
apps_df

App  ...      Installs
4534                                    Facebook  ...  1.000000e+09
9661                          WhatsApp Messenger  ...  1.000000e+09
5731                                   Instagram  ...  1.000000e+09
6546    Messenger – Text and Video Chat for Free  ...  1.000000e+09
2701                              Clash of Clans  ...  1.000000e+08
...                                          ...  ...           ...
4372                      FG Autumn Photo Puzzle  ...  1.000000e+01
4369  FE Other Disciplines Engineering Exam Prep  ...  1.000000e+02
9033                             Thyroid Nodules  ...  1.000000e+01
185                            ACCEPT CE MARKING  ...  1.000000e+01
5862           K-App Mitarbeiter Galeria Kaufhof  ...  1.000000e+02

[9648 rows x 13 columns]

In [22]:
# check on statistical information of the dataframe
apps_df.describe()

Reviews       Rating         Size        Price      Installs
count  9.648000e+03  9648.000000  9648.000000  9648.000000  9.648000e+03
mean   2.170487e+05     4.192485    17.820208     1.098122  7.806898e+06
std    1.832460e+06     0.496210    21.503151    16.861193  5.379975e+07
min    0.000000e+00     1.000000     0.000000     0.000000  0.000000e+00
25%    2.500000e+01     4.000000     2.900000     0.000000  1.000000e+03
50%    9.745000e+02     4.300000     9.200000     0.000000  1.000000e+05
75%    2.949750e+04     4.500000    25.000000     0.000000  1.000000e+06
max    7.815831e+07     5.000000   100.000000   400.000000  1.000000e+09

##Exploratory Analysis and Visualization
##Category 